src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,25 +1,163 @@
 package com.es.phoneshop.model.product;
 
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Currency;
 import java.util.List;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+
+    private long maxId;
+    private List<Product> products;
+    ReadWriteLock lock;
+
+    public ArrayListProductDao() {
+        this.products = new ArrayList<>();
+        lock = new ReentrantReadWriteLock();
+        saveSampleProducts();
+    }
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+
+        if (id == null) {
+            throw new IllegalArgumentException("Id can't be null");
+        }
+        lock.readLock().lock();
+        try {
+            return products.stream()
+                    .filter(it -> id.equals(it.getId()))
+                    .findAny()
+                    .orElse(null);
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     @Override
     public List<Product> findProducts() {
-        throw new RuntimeException("Not implemented");
+
+        lock.readLock().lock();
+        try {
+            return products.stream()
+                    .filter(it -> it.getPrice() != null)
+                    .filter(it -> it.getStock() > 0)
+                    .collect(Collectors.toList());
+        } finally {
+            lock.readLock().unlock();
+        }
     }
 
     @Override
-    public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+    public boolean save(Product product) {
+
+        if (product.getId() == null) {
+
+            lock.writeLock().lock();
+            try {
+                addProduct(product);
+                return true;
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+        } else {
+
+            Product existedProductWithThisId;
+
+            lock.readLock().lock();
+            try {
+                existedProductWithThisId = products.stream()
+                        .filter(it -> product.getId().equals(it.getId()))
+                        .findAny()
+                        .orElse(null);
+
+            } finally {
+                lock.readLock().unlock();
+            }
+
+            lock.writeLock().lock();
+            try {
+                if (existedProductWithThisId == null) {
+                    addProduct(product);
+                    return true;
+                } else {
+                    changeExistingProduct(existedProductWithThisId, product);
+                    return false;
+                }
+
+            } finally {
+                lock.writeLock().unlock();
+            }
```

Comment:
Can we do the following?

```
public boolean save(Product product) {
    **lock.writeLock().lock();**
    try {
        if (product.getId() == null) {
            **// No explicit locking here**
            addProduct(...)
        } else {
            **// No explicit locking here**
            getProduct(id)...
            changeExistingProduct(...)
            addProduct(...)
        }
    } catch(...) {
        ...
    } finally {
        **lock.writeLock().unlock();**
    }

}
```

This way we use the following approach:
lock writeLock
addProduct
unlock writeLock

getProduct() method contains read locking so it's not needed to lock the readLock in save method.

Reply:
done